File: login-with-sendseven/java/src/main/java/com/sendseven/oauth/service/OAuthService.java
package com.sendseven.oauth.service;

import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.jwk.JWK;
import com.nimbusds.jose.jwk.JWKSet;
import com.nimbusds.jose.jwk.RSAKey;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import com.sendseven.oauth.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.UriComponentsBuilder;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.text.ParseException;
import java.util.Base64;
import java.util.Date;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * OAuth2/OIDC Service for SendSeven authentication.
 *
 * Handles the complete OAuth2 Authorization Code flow with PKCE:
 * - PKCE code generation
 * - State and nonce generation
 * - Authorization URL building
 * - Token exchange
 * - ID token verification using JWKS
 * - Token refresh
 * - Token revocation
 */
@Service
public class OAuthService {

    private static final Logger logger = LoggerFactory.getLogger(OAuthService.class);
    private static final SecureRandom secureRandom = new SecureRandom();

    @Value("${sendseven.client-id}")
    private String clientId;

    @Value("${sendseven.client-secret}")
    private String clientSecret;

    @Value("${sendseven.api-url}")
    private String apiUrl;

    @Value("${sendseven.redirect-uri}")
    private String redirectUri;

    private final WebClient webClient;

    // JWKS cache
    private final Map<String, JWKSet> jwksCache = new ConcurrentHashMap<>();
    private long jwksCacheTime = 0;
    private static final long JWKS_CACHE_DURATION_MS = 3600 * 1000; // 1 hour

    // OIDC config cache
    private OIDCConfig oidcConfigCache;
    private long oidcConfigCacheTime = 0;
    private static final long OIDC_CONFIG_CACHE_DURATION_MS = 3600 * 1000; // 1 hour

    public OAuthService(WebClient.Builder webClientBuilder) {
        this.webClient = webClientBuilder.build();
    }

    // =========================================================================
    // PKCE Generation
    // =========================================================================

    /**
     * Generate a cryptographically random code verifier for PKCE.
     *
     * The code verifier is a random string between 43 and 128 characters,
     * using URL-safe base64 characters (A-Z, a-z, 0-9, -, _, ~, .).
     *
     * @return A random code verifier string
     */
    public String generateCodeVerifier() {
        byte[] randomBytes = new byte[64];
        secureRandom.nextBytes(randomBytes);
        String verifier = Base64.getUrlEncoder().withoutPadding().encodeToString(randomBytes);
        // Ensure length is between 43-128 characters
        return verifier.substring(0, Math.min(128, verifier.length()));
    }

    /**
     * Generate S256 code challenge from the code verifier.
     *
     * The code challenge is the Base64URL-encoded SHA-256 hash of the verifier.
     *
     * @param codeVerifier The PKCE code verifier
     * @return The S256 code challenge
     */
    public String generateCodeChallenge(String codeVerifier) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(codeVerifier.getBytes(StandardCharsets.US_ASCII));
            return Base64.getUrlEncoder().withoutPadding().encodeToString(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    // =========================================================================
    // State and Nonce Generation
    // =========================================================================

    /**
     * Generate a random state parameter for CSRF protection.
     *
     * @return A random state string (32+ bytes, URL-safe)
     */
    public String generateState() {
        byte[] randomBytes = new byte[32];
        secureRandom.nextBytes(randomBytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(randomBytes);
    }

    /**
     * Generate a random nonce for ID token replay protection.
     *
     * @return A random nonce string (32+ bytes, URL-safe)
     */
    public String generateNonce() {
        byte[] randomBytes = new byte[32];
        secureRandom.nextBytes(randomBytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(randomBytes);
    }

    // =========================================================================
    // OAuth Session Management
    // =========================================================================

    /**
     * Create a new OAuth session with PKCE, state, and nonce.
     *
     * @return A new OAuthSession with all security parameters
     */
    public OAuthSession createOAuthSession() {
        String codeVerifier = generateCodeVerifier();
        String state = generateState();
        String nonce = generateNonce();
        return new OAuthSession(codeVerifier, state, nonce);
    }

    // =========================================================================
    // Authorization URL
    // =========================================================================

    /**
     * Build the authorization URL to redirect the user to SendSeven.
     *
     * @param oauthSession The OAuth session containing PKCE and security params
     * @return The complete authorization URL
     */
    public String buildAuthorizationUrl(OAuthSession oauthSession) {
        String codeChallenge = generateCodeChallenge(oauthSession.getCodeVerifier());

        return UriComponentsBuilder.fromUriString(apiUrl)
                .path("/api/v1/oauth-apps/authorize")
                .queryParam("client_id", clientId)
                .queryParam("redirect_uri", redirectUri)
                .queryParam("response_type", "code")
                .queryParam("scope", "openid profile email offline_access")
                .queryParam("state", oauthSession.getState())
                .queryParam("code_challenge", codeChallenge)
                .queryParam("code_challenge_method", "S256")
                .queryParam("nonce", oauthSession.getNonce())
                .build()
                .toUriString();
    }

    // =========================================================================
    // OIDC Discovery
    // =========================================================================

    /**
     * Fetch the OIDC discovery document.
     *
     * Caches the result for 1 hour.
     *
     * @return The OIDC configuration
     */
    public OIDCConfig getOIDCConfig() {
        long now = System.currentTimeMillis();

        if (oidcConfigCache != null && (now - oidcConfigCacheTime) < OIDC_CONFIG_CACHE_DURATION_MS) {
            return oidcConfigCache;
        }

        String discoveryUrl = apiUrl + "/.well-known/openid-configuration";
        logger.info("Fetching OIDC discovery from: {}", discoveryUrl);

        try {
            OIDCConfig config = webClient.get()
                    .uri(discoveryUrl)
                    .retrieve()
                    .bodyToMono(OIDCConfig.class)
                    .block();

            oidcConfigCache = config;
            oidcConfigCacheTime = now;

            return config;
        } catch (Exception e) {
            logger.error("Failed to fetch OIDC configuration", e);
            throw new RuntimeException("Failed to fetch OIDC configuration: " + e.getMessage(), e);
        }
    }

    /**
     * Fetch the JSON Web Key Set for ID token verification.
     *
     * Caches the result for 1 hour.
     *
     * @param jwksUri The JWKS URI from OIDC discovery
     * @return The JWK Set
     */
    public JWKSet getJWKS(String jwksUri) {
        long now = System.currentTimeMillis();

        JWKSet cached = jwksCache.get(jwksUri);
        if (cached != null && (now - jwksCacheTime) < JWKS_CACHE_DURATION_MS) {
            return cached;
        }

        logger.info("Fetching JWKS from: {}", jwksUri);

        try {
            JWKSet jwkSet = JWKSet.load(new URL(jwksUri));
            jwksCache.put(jwksUri, jwkSet);
            jwksCacheTime = now;
            return jwkSet;
        } catch (Exception e) {
            logger.error("Failed to fetch JWKS", e);
            throw new RuntimeException("Failed to fetch JWKS: " + e.getMessage(), e);
        }
    }

    // =========================================================================
    // Token Exchange
    // =========================================================================

    /**
     * Exchange an authorization code for tokens.
     *
     * @param code The authorization code from the callback
     * @param codeVerifier The PKCE code verifier
     * @return The token response containing access_token, refresh_token, and id_token
     */
    public TokenResponse exchangeCodeForTokens(String code, String codeVerifier) {
        String tokenUrl = apiUrl + "/api/v1/oauth-apps/token";
        logger.info("Exchanging code for tokens at: {}", tokenUrl);

        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("grant_type", "authorization_code");
        formData.add("code", code);
        formData.add("client_id", clientId);
        formData.add("client_secret", clientSecret);
        formData.add("redirect_uri", redirectUri);
        formData.add("code_verifier", codeVerifier);

        try {
            TokenResponse response = webClient.post()
                    .uri(tokenUrl)
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(BodyInserters.fromFormData(formData))
                    .retrieve()
                    .bodyToMono(TokenResponse.class)
                    .block();

            logger.info("Token exchange successful");
            return response;
        } catch (Exception e) {
            logger.error("Token exchange failed", e);
            throw new RuntimeException("Failed to exchange code for tokens: " + e.getMessage(), e);
        }
    }

    // =========================================================================
    // ID Token Verification
    // =========================================================================

    /**
     * Verify an ID token and extract claims.
     *
     * Performs the following validations:
     * - Signature verification using JWKS
     * - Issuer (iss) matches expected
     * - Audience (aud) matches client_id
     * - Token is not expired (exp)
     * - Nonce matches the one sent in authorization request
     *
     * @param idToken The ID token JWT
     * @param expectedNonce The nonce that was sent in the authorization request
     * @return The verified ID token claims
     * @throws RuntimeException if verification fails
     */
    public IdTokenClaims verifyIdToken(String idToken, String expectedNonce) {
        try {
            // Parse the JWT
            SignedJWT signedJWT = SignedJWT.parse(idToken);

            // Get OIDC config for issuer and JWKS URI
            OIDCConfig oidcConfig = getOIDCConfig();

            // Fetch JWKS
            JWKSet jwkSet = getJWKS(oidcConfig.getJwksUri());

            // Get the key ID from the JWT header
            String kid = signedJWT.getHeader().getKeyID();
            JWSAlgorithm alg = signedJWT.getHeader().getAlgorithm();

            logger.info("Verifying ID token with kid: {}, alg: {}", kid, alg);

            // Find the matching key
            JWK jwk = jwkSet.getKeyByKeyId(kid);
            if (jwk == null) {
                throw new RuntimeException("No matching key found for kid: " + kid);
            }

            // Verify the signature
            if (!(jwk instanceof RSAKey)) {
                throw new RuntimeException("Expected RSA key but got: " + jwk.getKeyType());
            }

            RSAKey rsaKey = (RSAKey) jwk;
            com.nimbusds.jose.crypto.RSASSAVerifier verifier =
                    new com.nimbusds.jose.crypto.RSASSAVerifier(rsaKey);

            if (!signedJWT.verify(verifier)) {
                throw new RuntimeException("ID token signature verification failed");
            }

            logger.info("ID token signature verified successfully");

            // Extract and validate claims
            JWTClaimsSet claims = signedJWT.getJWTClaimsSet();

            // Validate issuer
            String issuer = claims.getIssuer();
            if (!oidcConfig.getIssuer().equals(issuer)) {
                throw new RuntimeException("Invalid issuer. Expected: " + oidcConfig.getIssuer() +
                        ", got: " + issuer);
            }

            // Validate audience
            if (!claims.getAudience().contains(clientId)) {
                throw new RuntimeException("Invalid audience. Expected: " + clientId +
                        ", got: " + claims.getAudience());
            }

            // Validate expiration
            Date exp = claims.getExpirationTime();
            if (exp == null || new Date().after(exp)) {
                throw new RuntimeException("ID token has expired");
            }

            // Validate nonce
            String nonce = claims.getStringClaim("nonce");
            if (expectedNonce != null && !expectedNonce.equals(nonce)) {
                throw new RuntimeException("Invalid nonce. Expected: " + expectedNonce +
                        ", got: " + nonce);
            }

            // Build claims object
            IdTokenClaims idTokenClaims = new IdTokenClaims();
            idTokenClaims.setIssuer(issuer);
            idTokenClaims.setSubject(claims.getSubject());
            idTokenClaims.setAudience(clientId);
            idTokenClaims.setExpirationTime(exp);
            idTokenClaims.setIssuedAt(claims.getIssueTime());
            idTokenClaims.setNonce(nonce);
            idTokenClaims.setEmail(claims.getStringClaim("email"));
            idTokenClaims.setEmailVerified(getBooleanClaim(claims, "email_verified"));
            idTokenClaims.setName(claims.getStringClaim("name"));
            idTokenClaims.setPicture(claims.getStringClaim("picture"));
            idTokenClaims.setTenantId(claims.getStringClaim("tenant_id"));

            logger.info("ID token claims verified: {}", idTokenClaims);

            return idTokenClaims;

        } catch (ParseException e) {
            logger.error("Failed to parse ID token", e);
            throw new RuntimeException("Invalid ID token format: " + e.getMessage(), e);
        } catch (Exception e) {
            logger.error("ID token verification failed", e);
            throw new RuntimeException("ID token verification failed: " + e.getMessage(), e);
        }
    }

    private Boolean getBooleanClaim(JWTClaimsSet claims, String claimName) {
        try {
            return claims.getBooleanClaim(claimName);
        } catch (ParseException e) {
            return null;
        }
    }

    // =========================================================================
    // User Info
    // =========================================================================

    /**
     * Fetch user information using the access token.
     *
     * @param accessToken The OAuth2 access token
     * @return The user info
     */
    public UserInfo getUserInfo(String accessToken) {
        String userinfoUrl = apiUrl + "/api/v1/oauth-apps/userinfo";
        logger.info("Fetching user info from: {}", userinfoUrl);

        try {
            UserInfo userInfo = webClient.get()
                    .uri(userinfoUrl)
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + accessToken)
                    .retrieve()
                    .bodyToMono(UserInfo.class)
                    .block();

            logger.info("User info fetched: {}", userInfo);
            return userInfo;
        } catch (Exception e) {
            logger.error("Failed to fetch user info", e);
            throw new RuntimeException("Failed to fetch user info: " + e.getMessage(), e);
        }
    }

    // =========================================================================
    // Token Refresh
    // =========================================================================

    /**
     * Refresh tokens using a refresh token.
     *
     * @param refreshToken The refresh token
     * @return New token response with fresh access token
     */
    public TokenResponse refreshTokens(String refreshToken) {
        String tokenUrl = apiUrl + "/api/v1/oauth-apps/token";
        logger.info("Refreshing tokens at: {}", tokenUrl);

        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("grant_type", "refresh_token");
        formData.add("refresh_token", refreshToken);
        formData.add("client_id", clientId);
        formData.add("client_secret", clientSecret);

        try {
            TokenResponse response = webClient.post()
                    .uri(tokenUrl)
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(BodyInserters.fromFormData(formData))
                    .retrieve()
                    .bodyToMono(TokenResponse.class)
                    .block();

            logger.info("Token refresh successful");
            return response;
        } catch (Exception e) {
            logger.error("Token refresh failed", e);
            throw new RuntimeException("Failed to refresh tokens: " + e.getMessage(), e);
        }
    }

    // =========================================================================
    // Token Revocation
    // =========================================================================

    /**
     * Revoke a token (access token or refresh token).
     *
     * @param token The token to revoke
     * @param tokenTypeHint Optional hint: "access_token" or "refresh_token"
     */
    public void revokeToken(String token, String tokenTypeHint) {
        String revokeUrl = apiUrl + "/api/v1/oauth-apps/revoke";
        logger.info("Revoking token at: {}", revokeUrl);

        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("token", token);
        if (tokenTypeHint != null) {
            formData.add("token_type_hint", tokenTypeHint);
        }
        formData.add("client_id", clientId);
        formData.add("client_secret", clientSecret);

        try {
            webClient.post()
                    .uri(revokeUrl)
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(BodyInserters.fromFormData(formData))
                    .retrieve()
                    .toBodilessEntity()
                    .block();

            logger.info("Token revoked successfully");
        } catch (Exception e) {
            // Token revocation errors are typically non-fatal
            logger.warn("Token revocation failed (continuing): {}", e.getMessage());
        }
    }
}

File: login-with-sendseven/csharp/Services/OAuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using LoginWithSendSeven.Models;
using Microsoft.IdentityModel.Tokens;

namespace LoginWithSendSeven.Services;

/// <summary>
/// Service for OAuth2/OIDC operations with SendSeven.
/// Implements the complete Authorization Code flow with PKCE.
/// </summary>
public class OAuthService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;
    private readonly ILogger<OAuthService> _logger;

    // OIDC configuration cache
    private OidcConfiguration? _oidcConfig;
    private DateTime _oidcConfigFetchedAt;
    private readonly TimeSpan _oidcConfigCacheDuration = TimeSpan.FromHours(1);

    // JWKS cache
    private JwksResponse? _jwks;
    private DateTime _jwksFetchedAt;
    private readonly TimeSpan _jwksCacheDuration = TimeSpan.FromHours(1);

    public OAuthService(HttpClient httpClient, IConfiguration configuration, ILogger<OAuthService> logger)
    {
        _httpClient = httpClient;
        _configuration = configuration;
        _logger = logger;
    }

    #region Configuration

    /// <summary>
    /// Get the SendSeven API base URL.
    /// </summary>
    public string ApiUrl => Environment.GetEnvironmentVariable("SENDSEVEN_API_URL")
        ?? _configuration["SendSeven:ApiUrl"]
        ?? "https://api.sendseven.com";

    /// <summary>
    /// Get the OAuth client ID.
    /// </summary>
    public string ClientId => Environment.GetEnvironmentVariable("SENDSEVEN_CLIENT_ID")
        ?? _configuration["SendSeven:ClientId"]
        ?? throw new InvalidOperationException("SENDSEVEN_CLIENT_ID is not configured");

    /// <summary>
    /// Get the OAuth client secret.
    /// </summary>
    public string ClientSecret => Environment.GetEnvironmentVariable("SENDSEVEN_CLIENT_SECRET")
        ?? _configuration["SendSeven:ClientSecret"]
        ?? throw new InvalidOperationException("SENDSEVEN_CLIENT_SECRET is not configured");

    /// <summary>
    /// Get the redirect URI.
    /// </summary>
    public string RedirectUri => Environment.GetEnvironmentVariable("REDIRECT_URI")
        ?? _configuration["SendSeven:RedirectUri"]
        ?? "http://localhost:3000/callback";

    /// <summary>
    /// Get the requested scopes.
    /// </summary>
    public string Scopes => _configuration["SendSeven:Scopes"]
        ?? "openid profile email offline_access";

    #endregion

    #region PKCE Generation

    /// <summary>
    /// Generate a cryptographically random code verifier for PKCE.
    /// Length should be between 43 and 128 characters.
    /// </summary>
    public string GenerateCodeVerifier(int length = 64)
    {
        // Ensure length is within spec (43-128)
        length = Math.Max(43, Math.Min(128, length));

        // Generate random bytes
        var bytes = new byte[length];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(bytes);

        // Convert to URL-safe base64
        return Base64UrlEncode(bytes).Substring(0, length);
    }

    /// <summary>
    /// Generate S256 code challenge from the verifier.
    /// SHA256 hash, base64url encoded, no padding.
    /// </summary>
    public string GenerateCodeChallenge(string verifier)
    {
        using var sha256 = SHA256.Create();
        var bytes = Encoding.UTF8.GetBytes(verifier);
        var hash = sha256.ComputeHash(bytes);
        return Base64UrlEncode(hash);
    }

    /// <summary>
    /// Generate a random state parameter for CSRF protection.
    /// </summary>
    public string GenerateState()
    {
        var bytes = new byte[32];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(bytes);
        return Base64UrlEncode(bytes);
    }

    /// <summary>
    /// Generate a random nonce for ID token replay protection.
    /// </summary>
    public string GenerateNonce()
    {
        var bytes = new byte[32];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(bytes);
        return Base64UrlEncode(bytes);
    }

    /// <summary>
    /// Generate complete OAuth state for the authorization flow.
    /// </summary>
    public OAuthState GenerateOAuthState()
    {
        var codeVerifier = GenerateCodeVerifier();
        return new OAuthState
        {
            State = GenerateState(),
            Nonce = GenerateNonce(),
            CodeVerifier = codeVerifier,
            CreatedAt = DateTime.UtcNow
        };
    }

    #endregion

    #region OIDC Discovery

    /// <summary>
    /// Fetch the OIDC discovery document.
    /// </summary>
    public async Task<OidcConfiguration> GetOidcConfigurationAsync()
    {
        // Return cached if still valid
        if (_oidcConfig != null && DateTime.UtcNow - _oidcConfigFetchedAt < _oidcConfigCacheDuration)
        {
            return _oidcConfig;
        }

        var discoveryUrl = $"{ApiUrl.TrimEnd('/')}/.well-known/openid-configuration";
        _logger.LogInformation("Fetching OIDC discovery document from {Url}", discoveryUrl);

        var response = await _httpClient.GetAsync(discoveryUrl);
        response.EnsureSuccessStatusCode();

        var content = await response.Content.ReadAsStringAsync();
        _oidcConfig = JsonSerializer.Deserialize<OidcConfiguration>(content)
            ?? throw new InvalidOperationException("Failed to parse OIDC configuration");
        _oidcConfigFetchedAt = DateTime.UtcNow;

        return _oidcConfig;
    }

    /// <summary>
    /// Fetch the JSON Web Key Set for ID token verification.
    /// </summary>
    public async Task<JwksResponse> GetJwksAsync()
    {
        // Return cached if still valid
        if (_jwks != null && DateTime.UtcNow - _jwksFetchedAt < _jwksCacheDuration)
        {
            return _jwks;
        }

        var oidcConfig = await GetOidcConfigurationAsync();
        _logger.LogInformation("Fetching JWKS from {Url}", oidcConfig.JwksUri);

        var response = await _httpClient.GetAsync(oidcConfig.JwksUri);
        response.EnsureSuccessStatusCode();

        var content = await response.Content.ReadAsStringAsync();
        _jwks = JsonSerializer.Deserialize<JwksResponse>(content)
            ?? throw new InvalidOperationException("Failed to parse JWKS");
        _jwksFetchedAt = DateTime.UtcNow;

        return _jwks;
    }

    #endregion

    #region Authorization

    /// <summary>
    /// Build the authorization URL for the OAuth2 flow.
    /// </summary>
    public string BuildAuthorizationUrl(OAuthState state)
    {
        var codeChallenge = GenerateCodeChallenge(state.CodeVerifier);

        var parameters = new Dictionary<string, string>
        {
            ["client_id"] = ClientId,
            ["redirect_uri"] = RedirectUri,
            ["response_type"] = "code",
            ["scope"] = Scopes,
            ["state"] = state.State,
            ["code_challenge"] = codeChallenge,
            ["code_challenge_method"] = "S256",
            ["nonce"] = state.Nonce
        };

        var queryString = string.Join("&",
            parameters.Select(p => $"{Uri.EscapeDataString(p.Key)}={Uri.EscapeDataString(p.Value)}"));

        var authUrl = $"{ApiUrl.TrimEnd('/')}/api/v1/oauth-apps/authorize?{queryString}";
        _logger.LogInformation("Built authorization URL: {Url}", authUrl);

        return authUrl;
    }

    #endregion

    #region Token Exchange

    /// <summary>
    /// Exchange authorization code for tokens.
    /// </summary>
    public async Task<TokenResponse> ExchangeCodeForTokensAsync(string code, string codeVerifier)
    {
        var tokenUrl = $"{ApiUrl.TrimEnd('/')}/api/v1/oauth-apps/token";
        _logger.LogInformation("Exchanging code for tokens at {Url}", tokenUrl);

        var requestData = new Dictionary<string, string>
        {
            ["grant_type"] = "authorization_code",
            ["code"] = code,
            ["client_id"] = ClientId,
            ["client_secret"] = ClientSecret,
            ["redirect_uri"] = RedirectUri,
            ["code_verifier"] = codeVerifier
        };

        var content = new FormUrlEncodedContent(requestData);
        var response = await _httpClient.PostAsync(tokenUrl, content);

        var responseContent = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            var error = JsonSerializer.Deserialize<OAuthErrorResponse>(responseContent);
            throw new OAuthException(
                error?.Error ?? "token_exchange_failed",
                error?.ErrorDescription ?? $"HTTP {response.StatusCode}: {responseContent}");
        }

        return JsonSerializer.Deserialize<TokenResponse>(responseContent)
            ?? throw new InvalidOperationException("Failed to parse token response");
    }

    /// <summary>
    /// Refresh an expired access token using a refresh token.
    /// </summary>
    public async Task<TokenResponse> RefreshTokenAsync(string refreshToken)
    {
        var tokenUrl = $"{ApiUrl.TrimEnd('/')}/api/v1/oauth-apps/token";
        _logger.LogInformation("Refreshing token at {Url}", tokenUrl);

        var requestData = new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refreshToken,
            ["client_id"] = ClientId,
            ["client_secret"] = ClientSecret
        };

        var content = new FormUrlEncodedContent(requestData);
        var response = await _httpClient.PostAsync(tokenUrl, content);

        var responseContent = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            var error = JsonSerializer.Deserialize<OAuthErrorResponse>(responseContent);
            throw new OAuthException(
                error?.Error ?? "refresh_failed",
                error?.ErrorDescription ?? $"HTTP {response.StatusCode}: {responseContent}");
        }

        return JsonSerializer.Deserialize<TokenResponse>(responseContent)
            ?? throw new InvalidOperationException("Failed to parse token response");
    }

    #endregion

    #region ID Token Verification

    /// <summary>
    /// Verify the ID token signature and claims.
    /// </summary>
    public async Task<JwtSecurityToken> VerifyIdTokenAsync(string idToken, string expectedNonce)
    {
        _logger.LogInformation("Verifying ID token");

        // Get OIDC configuration for issuer
        var oidcConfig = await GetOidcConfigurationAsync();

        // Get JWKS for public keys
        var jwks = await GetJwksAsync();

        // Parse the token to get the key ID
        var handler = new JwtSecurityTokenHandler();
        var jwtToken = handler.ReadJwtToken(idToken);
        var kid = jwtToken.Header.Kid;

        _logger.LogDebug("Looking for key with kid: {Kid}", kid);

        // Find the matching key
        var matchingKey = jwks.Keys.FirstOrDefault(k => k.KeyId == kid);
        if (matchingKey == null)
        {
            throw new SecurityTokenValidationException($"No matching key found for kid: {kid}");
        }

        // Build the RSA security key from JWK
        var rsaKey = BuildRsaSecurityKey(matchingKey);

        // Validation parameters
        var validationParameters = new TokenValidationParameters
        {
            ValidateIssuer = true,
            ValidIssuer = oidcConfig.Issuer,
            ValidateAudience = true,
            ValidAudience = ClientId,
            ValidateLifetime = true,
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = rsaKey,
            ClockSkew = TimeSpan.FromMinutes(5)
        };

        // Validate the token
        handler.ValidateToken(idToken, validationParameters, out var validatedToken);

        // Verify nonce
        var nonceClaim = jwtToken.Claims.FirstOrDefault(c => c.Type == "nonce")?.Value;
        if (nonceClaim != expectedNonce)
        {
            throw new SecurityTokenValidationException(
                $"Invalid nonce. Expected: {expectedNonce}, Got: {nonceClaim}");
        }

        _logger.LogInformation("ID token verified successfully for subject: {Sub}",
            jwtToken.Subject);

        return jwtToken;
    }

    /// <summary>
    /// Build an RSA security key from a JWK.
    /// </summary>
    private RsaSecurityKey BuildRsaSecurityKey(JsonWebKeyModel jwk)
    {
        // Decode the modulus and exponent from base64url
        var modulus = Base64UrlDecode(jwk.Modulus);
        var exponent = Base64UrlDecode(jwk.Exponent);

        // Create RSA parameters
        var rsaParams = new RSAParameters
        {
            Modulus = modulus,
            Exponent = exponent
        };

        // Create RSA key
        var rsa = RSA.Create();
        rsa.ImportParameters(rsaParams);

        return new RsaSecurityKey(rsa) { KeyId = jwk.KeyId };
    }

    #endregion

    #region User Info

    /// <summary>
    /// Get user information from the userinfo endpoint.
    /// </summary>
    public async Task<UserInfo> GetUserInfoAsync(string accessToken)
    {
        var userinfoUrl = $"{ApiUrl.TrimEnd('/')}/api/v1/oauth-apps/userinfo";
        _logger.LogInformation("Fetching user info from {Url}", userinfoUrl);

        using var request = new HttpRequestMessage(HttpMethod.Get, userinfoUrl);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);

        var response = await _httpClient.SendAsync(request);
        var responseContent = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new OAuthException(
                "userinfo_failed",
                $"Failed to fetch user info: HTTP {response.StatusCode}: {responseContent}");
        }

        return JsonSerializer.Deserialize<UserInfo>(responseContent)
            ?? throw new InvalidOperationException("Failed to parse user info response");
    }

    #endregion

    #region Token Revocation

    /// <summary>
    /// Revoke a token (logout).
    /// </summary>
    public async Task RevokeTokenAsync(string token, string tokenTypeHint = "refresh_token")
    {
        var revokeUrl = $"{ApiUrl.TrimEnd('/')}/api/v1/oauth-apps/revoke";
        _logger.LogInformation("Revoking token at {Url}", revokeUrl);

        var requestData = new Dictionary<string, string>
        {
            ["token"] = token,
            ["token_type_hint"] = tokenTypeHint,
            ["client_id"] = ClientId,
            ["client_secret"] = ClientSecret
        };

        var content = new FormUrlEncodedContent(requestData);

        try
        {
            var response = await _httpClient.PostAsync(revokeUrl, content);
            // Per RFC 7009, revocation endpoint should return 200 even for invalid tokens
            _logger.LogInformation("Token revoked successfully");
        }
        catch (Exception ex)
        {
            // Don't fail logout if revocation fails
            _logger.LogWarning(ex, "Token revocation failed (continuing with logout)");
        }
    }

    #endregion

    #region Helpers

    /// <summary>
    /// Base64url encode without padding.
    /// </summary>
    private static string Base64UrlEncode(byte[] data)
    {
        return Convert.ToBase64String(data)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }

    /// <summary>
    /// Base64url decode (handles missing padding).
    /// </summary>
    private static byte[] Base64UrlDecode(string input)
    {
        // Add padding if necessary
        var padded = input
            .Replace('-', '+')
            .Replace('_', '/');

        switch (padded.Length % 4)
        {
            case 2: padded += "=="; break;
            case 3: padded += "="; break;
        }

        return Convert.FromBase64String(padded);
    }

    #endregion
}

/// <summary>
/// Exception for OAuth-specific errors.
/// </summary>
public class OAuthException : Exception
{
    public string Error { get; }
    public string? ErrorDescription { get; }

    public OAuthException(string error, string? errorDescription = null)
        : base(errorDescription ?? error)
    {
        Error = error;
        ErrorDescription = errorDescription;
    }
}
